### PR TITLE
Api-175: category integration tests

### PR DESCRIPTION
**Pull request by LaurentPetard**

### CODE REVIEW

File: tests/Api/ApiTestCase.php [hunk 1]
```diff
@@ -80,19 +80,37 @@ protected function getConfiguration()
...
+        $actualContent = $this->sortResourceContent($actualContent);
+
+        $expectedContent = array_merge($actualContent, $expectedContent);
+
+        $this->assertSame($expectedContent, $actualContent);
+    }
+
+    /**
+     * Sort a resource content to be able to compare it with another one.
+     * The order of elements in an associative arrays is important in PHPUnit but not for us.
```
Comment (ahocquard) [inline tests/Api/ApiTestCase.php]: associative array (singular)

File: tests/Api/ApiTestCase.php [hunk 2]
```diff
@@ -103,6 +121,7 @@ public function assertSameContent(array $expectedContent, array $actualContent)
      */
     protected function computeArrayDiff(array $expectedValues, array $actualValues)
     {
+
```
Comment (ahocquard) [inline tests/Api/ApiTestCase.php]: blank line

File: tests/Api/ApiTestCase.php [hunk 3]
```diff
@@ -103,6 +121,7 @@ public function assertSameContent(array $expectedContent, array $actualContent)
      */
     protected function computeArrayDiff(array $expectedValues, array $actualValues)
```
Comment (ahocquard) [inline tests/Api/ApiTestCase.php]: could be delete now ?

File: tests/Api/CategoryApiIntegration.php [hunk 4]
```diff
@@ -0,0 +1,237 @@
+<?php
+
+namespace Akeneo\Pim\tests\Api;
+
+
+use Akeneo\Pim\Api\CategoryApi;
+
+class CategoryApiIntegration extends ApiTestCase
```
Comment (ahocquard) [inline tests/Api/CategoryApiIntegration.php]: Maybe better to split it as for the PIM: one class for each method (you can put listPerPage and all in the same file). We will probably enrich the tests over the time (new tests or new methods/endpoints).

File: tests/Api/CategoryApiIntegration.php [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            'labels' => [
+                'en_US' => 'The Pumps',
+                'fr_FR' => 'Les Pumps',
+            ],
+        ], $category);
+    }
+
+    public function testUpsert()
+    {
+        $api = $this->createClient()->getCategoryApi();
```
Comment (ahocquard) [inline tests/Api/CategoryApiIntegration.php]: Better to split it in two distinct tests

File: tests/Api/ApiTestCase.php [hunk 6]
```diff
@@ -80,19 +80,37 @@ protected function getConfiguration()
...
+     * Sort a resource content to be able to compare it with another one.
+     * The order of elements in an associative arrays is important in PHPUnit but not for us.
+     * So we force the order of the associative arrays to be identical to be able to use them in a PHPUnit assertion.
+     * This sort has no consequences for sequential arrays with numeric keys.
+     *
+     * @param array $resourceContent
+     *
+     * @return array Sorted resource content
+     */
+    private function sortResourceContent(array $resourceContent)
```
Comment (ahocquard) [inline tests/Api/ApiTestCase.php]: Much better this way 👍

File: tests/Api/ApiTestCase.php [hunk 7]
```diff
@@ -80,19 +80,37 @@ protected function getConfiguration()
...
-                'Failed asserting that the content has the expected body.'
-                .PHP_EOL
-                .'Differences between expected and actual content :'
-                .PHP_EOL
-                .var_export($differences, true)
-            );
+        $expectedContent = $this->sortResourceContent($expectedContent);
+        $actualContent = $this->sortResourceContent($actualContent);
+
+        $expectedContent = array_merge($actualContent, $expectedContent);
```
Comment (ahocquard) [inline tests/Api/ApiTestCase.php]: array_merge_recursive

File: tests/Api/ApiTestCase.php [hunk 8]
```diff
@@ -75,53 +75,83 @@ protected function getConfiguration()
...
+        foreach ($expectedContent as $key => $value) {
+            if (is_array($value) && isset($actualContent[$key]) && is_array($actualContent[$key])) {
+                $expectedContent[$key] = $this->mergeResourceContents($actualContent[$key], $expectedContent[$key]);
+            }
+        }
+
+        if ($this->isAssociativeArray($expectedContent)) {
+            $mergedContent = array_merge($actualContent, $expectedContent);
+        } else {
+            $mergedContent = $expectedContent;
```
Comment (ahocquard) [inline tests/Api/ApiTestCase.php]: `$mergedContent = $actualContent` ?

File: tests/Api/Category/CreateCategoryIntegration.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        $this->assertSameContent([
+            'code'   => 'pumps',
+            'parent' => '2014_collection',
+            'labels' => [
+                'en_US' => 'The Pumps',
+                'fr_FR' => 'Les Pumps',
+            ],
+        ], $category);
+    }
+
```
Comment (ahocquard) [inline tests/Api/Category/CreateCategoryIntegration.php]: blank line.

Test exception with an already existing category as well.

File: tests/Api/Category/ListCategoryIntegration.php [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+        ]), $secondPage->getPreviousLink());
+
+        $categories = $secondPage->getItems();
+        $this->assertCount(2 ,$categories);
+        $this->assertSameContent($expectedCategories[2], $categories[0]);
+        $this->assertSameContent($expectedCategories[3], $categories[1]);
+
+        $lastPage = $secondPage->getNextPage();
+        $this->assertInstanceOf(PageInterface::class, $lastPage);
+        $this->assertNull($lastPage->getNextPage());
```
Comment (ahocquard) [inline tests/Api/Category/ListCategoryIntegration.php]: test `hasNextPage()` as well pls

File: tests/Api/Category/ListCategoryIntegration.php [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    {
+        $api = $this->createClient()->getCategoryApi();
+        $uriGenerator = $this->getUriGenerator();
+        $expectedCategories = $this->getExpectedCategories();
+
+        $firstPage = $api->listPerPage(2);
+        $this->assertInstanceOf(PageInterface::class, $firstPage);
+        $this->assertNull($firstPage->getCount());
+        $this->assertNull($firstPage->getPreviousLink());
+        $this->assertNull($firstPage->getPreviousPage());
```
Comment (ahocquard) [inline tests/Api/Category/ListCategoryIntegration.php]: add a test on `hasPreviousPage()`

File: tests/Api/Category/UpsertCategoryIntegration.php [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        $category = $api->get('booties');
+        $this->assertSameContent([
+            'code' => 'booties',
+            'parent' => 'summer_collection',
+            'labels' => [
+                'en_US' => 'Booties',
+                'fr_FR' => 'Bottines',
+            ],
+        ], $category);
+    }
```
Comment (ahocquard) [inline tests/Api/Category/UpsertCategoryIntegration.php]: test failing upsert as well: problem on validation in the updater (wrong data type for example) and a problem in the Symfony validation.
This way, we have 1 test on each main exception.

File: tests/Api/Category/UpsertListCategoryIntegration.php [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+            'code'        => 'sandals',
+            'status_code' => 204,
+        ], $responseLines[1]);
+
+        $this->assertSame([
+            'line'        => 2,
+            'code'        => 'booties',
+            'status_code' => 201,
+        ], $responseLines[2]);
+    }
```
Comment (ahocquard) [inline tests/Api/Category/UpsertListCategoryIntegration.php]: you can add a line without the code in the request, and a line too long as well.

File: tests/Api/ApiTestCase.php (hunk 8 above)
Comment (LaurentPetard) [inline tests/Api/ApiTestCase.php]: Nope. For a numeric keys array we have to keep the expected one intact. The merged content is the one who will compared with the actual.

File: tests/Api/Category/CreateCategoryIntegration.php (hunk 9 above)
Comment (skeleton) [inline tests/Api/Category/CreateCategoryIntegration.php]: plus add a fail on creation please

File: tests/Api/Category/GetCategoryIntegration.php [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+class GetCategoryIntegration extends ApiTestCase
+{
+    public function testGet()
+    {
+        $api = $this->createClient()->getCategoryApi();
+
+        $category = $api->get('sandals');
+
+        $this->assertInternalType('array', $category);
```
Comment (skeleton) [inline tests/Api/Category/GetCategoryIntegration.php]: not sure this test is relevant as you test the full response just below

File: tests/Api/Category/ListCategoryIntegration.php (hunk 11 above)
Comment (skeleton) [inline tests/Api/Category/ListCategoryIntegration.php]: and `$this->assertFalse($firstPage->hasPreviousPage());`, etc

File: tests/Api/Category/ListCategoryIntegration.php [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        ]), $firstPage->getNextLink());
+
+        $categories = $firstPage->getItems();
+        $this->assertCount(2 ,$categories);
+        $this->assertSameContent($expectedCategories[0], $categories[0]);
+        $this->assertSameContent($expectedCategories[1], $categories[1]);
+
+        $secondPage = $firstPage->getNextPage();
+        $this->assertInstanceOf(PageInterface::class, $secondPage);
+        $this->assertSame($uriGenerator->generate(CategoryApi::CATEGORIES_PATH, [], [
```
Comment (skeleton) [inline tests/Api/Category/ListCategoryIntegration.php]: not a big fan of this kind of test. If we introduce something wrong/useless/regression inside uriGenerator, your test will always be OK. I prefer hardcode url here to have a comparison like that:
`$this->assertSame('http://localhost/api/rest/v1/categories?page=1&limit=2&with_count=false', $secondPage->getPreviousLink())`

File: tests/Api/Category/ListCategoryIntegration.php [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+
+    protected function getExpectedCategories()
+    {
+        $uriGenerator = $this->getUriGenerator();
+
+        return [
+            [
+                '_links' => [
+                    'self' => [
+                        'href' => $uriGenerator->generate(CategoryApi::CATEGORY_PATH, ['code' => '2014_collection']),
```
Comment (skeleton) [inline tests/Api/Category/ListCategoryIntegration.php]: same comment as above, avoid to call service to generate your route. hardcode them please

File: tests/Api/Category/ListCategoryIntegration.php [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+        foreach ($categories as $key => $category) {
+            $this->assertInternalType('array', $category);
+            $this->assertSameContent($expectedCategories[$key], $category);
+            $categoriesCount++;
+        }
+
+        $this->assertSame(5, $categoriesCount);
+    }
+
+    protected function getExpectedCategories()
```
Comment (skeleton) [inline tests/Api/Category/ListCategoryIntegration.php]: phpdoc

File: tests/Api/Category/ListCategoryIntegration.php [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+    public function testListPerPageWithCount()
+    {
+        $api = $this->createClient()->getCategoryApi();
+
+        $firstPage = $api->listPerPage(2, true);
+        $this->assertInstanceOf(PageInterface::class, $firstPage);
+        $this->assertSame(5, $firstPage->getCount());
+    }
+
+    public function testAll()
```
Comment (skeleton) [inline tests/Api/Category/ListCategoryIntegration.php]: please add a test where $queryParamers in `listPerPage()` & `all()` to check if parameters are well injected in URI

File: tests/Api/Category/GetCategoryIntegration.php (hunk 14 above)
Comment (LaurentPetard) [inline tests/Api/Category/GetCategoryIntegration.php]: It's mostly to have a assertion fail instead of a php error if the response is not an array.

Comment (skeleton) [inline tests/Api/Category/GetCategoryIntegration.php]: ok, but why only on this test ?

**Pull request merged**